### ISSUE TITLE
Bundler settings might allow users to execute code on the web server

### ISSUE THREAD

**Issue by CSC-swesters** [OSC/ondemand#1749]
Although the issue subject sounds alarming, we realize that OOD has the usr apps, which allow users to run code while developing apps. We also note, that when code executes, it is run as the non-privileged user. However, all sites might not wish to allow users to execute their own code on the web server.

We noticed that with OOD version 2.0.20 (RHEL7 RPM installation), the default Bundler configuration will allow users to configure their Ruby environments to load gems from a path (e.g. their home directories) that overrides the settings used by the OOD apps, e.g. the Dashboard and Myjobs apps. This seems to be due to the lack of an explicit gem path setting in the local app configuration, see the [bundler documentation](https://bundler.io/man/bundle-config.1.html) regarding the configuration loading order, if I have understood it correctly.

We found that this behavior could be prevented by running the following command inside the app directory (e.g. `apps/sys/dashboard/`):

```
bin/bundle config set --local path.system true
```

This will set `BUNDLE_PATH__SYSTEM: "true"` in the `<app dir>/.bundle/config` file.

Is this something which could be set by default in the upstream, in order to make the default installation more secure? Thanks in advance!

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1202821104799596/1202823687202198) by [Unito](https://www.unito.io)

**Comment by johrstrom** [comment]
There are ways to set the environment variables for the entire PUN, but as you indicate that may throw off development.

I think you may be right, that we'd just want to just supply the `.bundle/config` file with our system apps, though when we're developing them somehow turn that off or allow for changes? 

I believe we're those files are in the gitignore right now, so maybe we could add them as a part of the packaging.

**Comment by CSC-swesters** [comment]
I don't have a development setup of OOD personally, so I cannot comment on how it would technically work out between development and operations. My opinion is that it would be fair to provide production configurations with the RPM packages, and supply instructions for how development is enabled, if there is a need for further tweaks like in this case.

Did I understand correctly that the `BUNDLE_PATH__SYSTEM="true"` configuration could be delivered as a PUN environment variable, effectively applying to all ruby applications within OOD? If that's the case, perhaps that should be the suggested default, which is easy to reconfigure for development instances?

**Comment by johrstrom** [comment]
> Did I understand correctly that the BUNDLE_PATH__SYSTEM="true" configuration could be delivered as a PUN environment variable, effectively applying to all ruby applications within OOD?

Yes and you can do this now by using the `pun_custom_env` variable in nginx_stage.yml to 

```
pun_custom_env:
  BUNDLE_PATH__SYSTEM: true
```

You've submitted patches before - are you developing them through `/var/www/sys/`?  https://github.com/OSC/ondemand/blob/master/apps/dashboard/README.md

**Comment by johrstrom** [comment]
The documentation says that this `BUNDLE_PATH__SYSTEM` is actually where bundle will try to _install_. Seems like If we really want to force that path, we should use `path` because that will override GEM_HOME and GEM_PATH.

But we're already setting `GEM_PATH` elsewhere so those 2 strategies conflict right now (especially with how we install gems and have many directories for them).

@treydock do you have any thoughts?

**Comment by CSC-swesters** [comment]
Thank you for your suggestions! We'll test them out and see what works best.

We have not been using the `usr` apps, but instead we have a couple of separate instances for internal development and testing, versus production. So yes, our developers are developing in the `sys` apps. And thanks to this fact, the PUN variable solution would be ideal to us, allowing development to continue internally, but disallowing Gem loading from untrusted sources in production.

Other sites might still be interested in an alternative solution of course, where the upstream `sys` apps could be secured from this issue by default. Thanks for the active discussion!

**Comment by johrstrom** [comment]
Cool. Yea I would look at these instructions here so you can develop without privilege escalation. These are instructions on how to develop the dashboard it'self, but once you enable development for a user they can then develop batch connect apps too in their own `~/ondemand/dev` directory.

 https://github.com/OSC/ondemand/blob/master/apps/dashboard/README.md

**Comment by CSC-swesters** [comment]
We tried removing the `.bundle/config` setting and using the `pun_custom_env` setting instead, and it looks like the effect is the same, which is good.

We also noticed an option in [your nginx_stage.yml example](https://github.com/OSC/ondemand/blob/master/nginx_stage/share/nginx_stage_example.yml#L240), which we had overlooked earlier. `disable_bundle_user_config: true` seems to be even more effective in limiting user overrides, by preventing more than the PATHs to change. I found it in the [pun.conf.erb](https://github.com/OSC/ondemand/blob/master/nginx_stage/templates/pun.conf.erb#L32) and it apparently sets `BUNDLE_USER_CONFIG` to `/dev/null`. We tested that this also works for us.

Based on the existence of the `disable_bundle_user_config` option, this issue could be closed, I guess. Unless you'd like to consider changing the default value for the variable, that is? Of course, that might be a breaking change for existing installations, so it needs some extra consideration. Perhaps for version 2.1.0?

**Comment by johrstrom** [comment]
I did look that up earlier in this thread, but the default is `true`. Meaning we're likely already setting BUNDLE_USER_CONFIG to /dev/null.  Or are you saying you set this to `false` and that's the value you're looking for?

I think the issue is that `GEM_HOME` (having gems in `~/.gems/`) is still being respected in the absence of bundler settings that would otherwise force a different path(s).

**Comment by CSC-swesters** [comment]
Sorry for the confusion, you're right. The default is `true`.

Our earlier testing of the PUNs was perhaps not fully cleaned from the earlier configuration attempts. I guess we'll revert back to the `BUNDLE_PATH__SYSTEM=true` alternative, which we originally found worked.

**Comment by robinkar** [comment]
`disable_bundler_user_config` set to `true` should have solved this issue, but it seems like there is a bug in Bundler causing `BUNDLE_USER_CONFIG` to have no effect, https://github.com/rubygems/rubygems/issues/4793.
In our environment the Bundler version is 2.2.24 (from Software Collections) and running `bundle config list` shows that OOD indeed takes configuration also from `~/.bundle/config`.

**Comment by johrstrom** [comment]
Sorry for not getting to this quicker.  I believe this is now fixed in every bundler version 2.3.0 and up. 

Going through all the distros we currently support, every bundler should have this fix. 

Ubuntu 22.04: 2.3.5
Ubuntu 24.04: 2.4.20
Debian 12: 2.3.15
EL8: 2.5.22
EL9: 2.5.22
EL10: 2.5.22
amazon linux 2023: 2.4.19
